### DISCUSSION TITLE
Accessing headers with the axum service

### DISCUSSION THREAD

**Discussion by musjj** [async-graphql/async-graphql#1599] [General]
The official example recommends something like this:

[`examples/axum/token-from-header/src/main.rs`](https://github.com/async-graphql/examples/blob/629f1b7113598b26940836b2cc2710de5100ad36/axum/token-from-header/src/main.rs)

```rust
async fn graphql_handler(
    State(schema): State<TokenSchema>,
    headers: HeaderMap,
    req: GraphQLRequest,
) -> GraphQLResponse {
    let mut req = req.into_inner();
    if let Some(token) = get_token_from_headers(&headers) {
        req = req.data(token);
    }
    schema.execute(req).await.into()
}
```

But you lose all the nice features/logic you would get with the `async_graphql_axum::GraphQL` service:

https://github.com/async-graphql/async-graphql/blob/3046ae7f06ac08a9b912d8655d17f5c7f8c663c0/integrations/axum/src/query.rs#L54-L88

Is there really no better way of accessing the headers with the axum service without copy-pasting all of the code above?

**Comment by janos-r** [comment]
I absolutely agree, I also have to do this manual execution that feels almost like a hack, totally skipping the intended API.

https://github.com/janos-r/axum-template/blob/4e95a988408a6770f62ea50bcae4d57074bfeb27/src/graphql.rs#L19-L46

But I do have an idea! If the official code in
https://github.com/async-graphql/async-graphql/blob/3046ae7f06ac08a9b912d8655d17f5c7f8c663c0/integrations/axum/src/query.rs#L39-L46

Wouldn't return an HTTP response but a `async_graphql::Response` simply just not doing the `.into_response()` at the end which shouldn't be necessary anyway I think, we than could call the function as it is for execution and than edit the result as we need, and later just return the same thing (and axum calls into_response on it) or we call into_response at the end in our handler ourselves.
That way we don't have to do the manual execution!
Also can someone explain to me why it is a tower_service in the first place? If this is supposed to be an axum integration, wouldn't it be better to have it as a simple function handler to give to a .post(gql_handler) directly?

Any ideas?

**Comment by janos-r** [comment]
> no better way of accessing the headers

In my case, I need to access the errors that are created in the `async_graphql::Response`, that's why I recommend (to the library authors) extracting the call function out of the tower_service and simply provide one recommended unified way how to execute the schema, while being able to edit the result afterwards.
But in your case, the headers are already there before the execution right? Can't you just make a middleware that you put in a layer before the gql? That would have access to get the headers and can pass something on in an Extension.
